File: version4j/src/test/java/com/only4/algorithm/version4j/leetcode/FindMedianSortedArraysTest.java
package com.only4.algorithm.version4j.leetcode;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class FindMedianSortedArraysTest {

    private final double DELTA = 0.00001;

    @Test
    public void testBothArraysNonEmptyEvenLength() {
        FindMedianSortedArrays solution = new FindMedianSortedArrays();
        int[] nums1 = {1, 3};
        int[] nums2 = {2, 4};
        // 合并后为 [1, 2, 3, 4]，中位数为 (2 + 3) / 2 = 2.5
        assertEquals(2.5, solution.findMedianSortedArrays(nums1, nums2), DELTA);
    }

    @Test
    public void testBothArraysNonEmptyOddLength() {
        FindMedianSortedArrays solution = new FindMedianSortedArrays();
        int[] nums1 = {1, 3};
        int[] nums2 = {2};
        // 合并后为 [1, 2, 3]，中位数为 2
        assertEquals(2.0, solution.findMedianSortedArrays(nums1, nums2), DELTA);
    }

    @Test
    public void testOneEmptyArray() {
        FindMedianSortedArrays solution = new FindMedianSortedArrays();
        int[] nums1 = {};
        int[] nums2 = {1, 2, 3, 4, 5};
        // 合并后为 [1, 2, 3, 4, 5]，中位数为 3
        assertEquals(3.0, solution.findMedianSortedArrays(nums1, nums2), DELTA);
    }

    @Test
    public void testEqualLengthArrays() {
        FindMedianSortedArrays solution = new FindMedianSortedArrays();
        int[] nums1 = {1, 3, 5, 7};
        int[] nums2 = {2, 4, 6, 8};
        // 合并后为 [1, 2, 3, 4, 5, 6, 7, 8]，中位数为 (4 + 5) / 2 = 4.5
        assertEquals(4.5, solution.findMedianSortedArrays(nums1, nums2), DELTA);
    }

    @Test
    public void testDifferentLengthArrays() {
        FindMedianSortedArrays solution = new FindMedianSortedArrays();
        int[] nums1 = {1, 2};
        int[] nums2 = {3, 4, 5, 6, 7, 8, 9};
        // 合并后为 [1, 2, 3, 4, 5, 6, 7, 8, 9]，中位数为 5
        assertEquals(5.0, solution.findMedianSortedArrays(nums1, nums2), DELTA);
    }

    @Test
    public void testWithLargeValues() {
        FindMedianSortedArrays solution = new FindMedianSortedArrays();
        int[] nums1 = {100000};
        int[] nums2 = {100001};
        // 合并后为 [100000, 100001]，中位数为 (100000 + 100001) / 2 = 100000.5
        assertEquals(100000.5, solution.findMedianSortedArrays(nums1, nums2), DELTA);
    }

    @Test
    public void testWithNegativeValues() {
        FindMedianSortedArrays solution = new FindMedianSortedArrays();
        int[] nums1 = {-10, -5, 0, 5};
        int[] nums2 = {-20, -15, 10, 15};
        // 合并后为 [-20, -15, -10, -5, 0, 5, 10, 15]，中位数为 (-5 + 0) / 2 = -2.5
        assertEquals(-2.5, solution.findMedianSortedArrays(nums1, nums2), DELTA);
    }

    @Test
    public void testFirstArrayLargerThanSecond() {
        FindMedianSortedArrays solution = new FindMedianSortedArrays();
        int[] nums1 = {1, 2, 3, 4, 5, 6};
        int[] nums2 = {7, 8};
        // 合并后为 [1, 2, 3, 4, 5, 6, 7, 8]，中位数为 (4 + 5) / 2 = 4.5
        assertEquals(4.5, solution.findMedianSortedArrays(nums1, nums2), DELTA);
    }

    @Test
    public void testSecondArrayLargerThanFirst() {
        FindMedianSortedArrays solution = new FindMedianSortedArrays();
        int[] nums1 = {7, 8};
        int[] nums2 = {1, 2, 3, 4, 5, 6};
        // 合并后为 [1, 2, 3, 4, 5, 6, 7, 8]，中位数为 (4 + 5) / 2 = 4.5
        assertEquals(4.5, solution.findMedianSortedArrays(nums1, nums2), DELTA);
    }

    @Test
    public void testOverlappingArrays() {
        FindMedianSortedArrays solution = new FindMedianSortedArrays();
        int[] nums1 = {1, 3, 5, 7};
        int[] nums2 = {2, 3, 5, 7};
        // 合并后为 [1, 2, 3, 3, 5, 5, 7, 7]，中位数为 (3 + 5) / 2 = 4
        assertEquals(4.0, solution.findMedianSortedArrays(nums1, nums2), DELTA);
    }

    @Test
    public void testSingleElementArrays() {
        FindMedianSortedArrays solution = new FindMedianSortedArrays();
        int[] nums1 = {1};
        int[] nums2 = {2};
        // 合并后为 [1, 2]，中位数为 (1 + 2) / 2 = 1.5
        assertEquals(1.5, solution.findMedianSortedArrays(nums1, nums2), DELTA);
    }
}

File: core/src/test/kotlin/com/only4/algorithm/leetcode/FindMedianSortedArraysTest.kt
package com.only4.algorithm.leetcode

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.util.stream.Stream

class FindMedianSortedArraysTest {

    @Test
    fun `测试正常情况`() {
        // 测试用例1：两个数组长度相等，合并后为偶数长度
        val nums1 = intArrayOf(1, 3)
        val nums2 = intArrayOf(2, 4)
        assertEquals(2.5, findMedianSortedArrays(nums1, nums2))

        // 测试用例2：两个数组长度不等，合并后为奇数长度
        val nums3 = intArrayOf(1, 3)
        val nums4 = intArrayOf(2)
        assertEquals(2.0, findMedianSortedArrays(nums3, nums4))

        // 测试用例3：两个数组长度不等，合并后为偶数长度
        val nums5 = intArrayOf(1, 2)
        val nums6 = intArrayOf(3, 4, 5, 6)
        assertEquals(3.5, findMedianSortedArrays(nums5, nums6))
    }

    @Test
    fun `测试边界情况`() {
        // 测试用例1：一个数组为空，另一个数组长度为奇数
        val nums1 = intArrayOf()
        val nums2 = intArrayOf(1, 2, 3)
        assertEquals(2.0, findMedianSortedArrays(nums1, nums2))

        // 测试用例2：一个数组为空，另一个数组长度为偶数
        val nums3 = intArrayOf(1, 2, 3, 4)
        val nums4 = intArrayOf()
        assertEquals(2.5, findMedianSortedArrays(nums3, nums4))

        // 测试用例3：两个数组都只有一个元素
        val nums5 = intArrayOf(1)
        val nums6 = intArrayOf(2)
        assertEquals(1.5, findMedianSortedArrays(nums5, nums6))
    }

    @Test
    fun `测试特殊情况`() {
        // 测试用例1：两个数组有重叠元素
        val nums1 = intArrayOf(1, 2, 3)
        val nums2 = intArrayOf(2, 3, 4)
        assertEquals(2.5, findMedianSortedArrays(nums1, nums2))

        // 测试用例2：一个数组的所有元素都小于另一个数组
        val nums3 = intArrayOf(1, 2, 3)
        val nums4 = intArrayOf(4, 5, 6)
        assertEquals(3.5, findMedianSortedArrays(nums3, nums4))

        // 测试用例3：两个数组完全相同
        val nums5 = intArrayOf(1, 2, 3)
        val nums6 = intArrayOf(1, 2, 3)
        assertEquals(2.0, findMedianSortedArrays(nums5, nums6))
    }

    @ParameterizedTest
    @MethodSource("provideTestCases")
    fun `参数化测试`(nums1: IntArray, nums2: IntArray, expected: Double) {
        assertEquals(expected, findMedianSortedArrays(nums1, nums2), 0.00001)
    }

    companion object {
        @JvmStatic
        fun provideTestCases(): Stream<Arguments> {
            return Stream.of(
                Arguments.of(intArrayOf(1, 3), intArrayOf(2), 2.0),
                Arguments.of(intArrayOf(1, 2), intArrayOf(3, 4), 2.5),
                Arguments.of(intArrayOf(), intArrayOf(1), 1.0),
                Arguments.of(intArrayOf(2), intArrayOf(), 2.0),
                Arguments.of(intArrayOf(1, 3, 5, 7), intArrayOf(2, 4, 6, 8), 4.5),
                Arguments.of(intArrayOf(1, 2, 3, 4, 5), intArrayOf(6, 7, 8, 9, 10), 5.5),
                Arguments.of(intArrayOf(1), intArrayOf(1), 1.0),
                Arguments.of(intArrayOf(1, 1, 1), intArrayOf(1, 1, 1), 1.0),
                Arguments.of(intArrayOf(1, 2), intArrayOf(1, 2), 1.5),
                Arguments.of(intArrayOf(1, 3, 5, 7, 9), intArrayOf(2, 4, 6, 8, 10), 5.5)
            )
        }
    }
} 